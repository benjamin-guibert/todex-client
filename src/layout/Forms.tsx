import React, { FC, useContext, useEffect, useState } from 'react'
import DepositWithdrawInput from 'components/DepositWithdrawInput'
import { BigNumber, constants } from 'ethers'
import { ExchangeContext } from './ExchangeContext'
import StyledForms from './Forms.style'
import OrderForm from 'components/OrderForm'
import Card from 'react-bootstrap/Card'
import Order from 'models/Order'
import OrdersList from 'components/OrdersList'
import { MetaMaskContext } from './MetaMaskContext'
import Button from 'react-bootstrap/Button'

const { Zero } = constants

const Forms: FC = () => {
  const { account } = useContext(MetaMaskContext)
  const {
    depositEther,
    depositToken,
    withdrawEther,
    withdrawToken,
    getTokenAllowance,
    approveToken,
    cancelOrder,
    ethBalance,
    tokenBalance,
    orders,
  } = useContext(ExchangeContext)
  const [userOrders, setUserOrders] = useState<Order[]>([])

  const isTokenDepositAllowed = async (amount: BigNumber) => {
    const allowance = await getTokenAllowance()

    return !!allowance?.gte(amount)
  }

  const DeleteButton: FC<{ order: Order }> = ({ order }) => {
    return (
      <Button variant="outline-secondary" size="sm" onClick={() => cancelOrder(order.id as string)}>
        Cancel
      </Button>
    )
  }

  useEffect(() => {
    setUserOrders(orders.filter(({ account: orderAccount }) => orderAccount === account))
  }, [account, orders])

  return (
    <StyledForms>
      <DepositWithdrawInput
        name="ETH"
        onDeposit={depositEther}
        currentBalance={ethBalance ? BigNumber.from(ethBalance) : Zero}
        onWithdraw={withdrawEther}
      />
      <DepositWithdrawInput
        name="TDX"
        isDepositAllowed={isTokenDepositAllowed}
        currentBalance={tokenBalance ? BigNumber.from(tokenBalance) : Zero}
        onDeposit={depositToken}
        onWithdraw={withdrawToken}
        onApprove={approveToken}
      />
      <OrderForm />
      {!!userOrders.length && (
        <Card className="m-3 px-2 py-1">
          <OrdersList orders={userOrders} actions={[DeleteButton]} styleType />
        </Card>
      )}
    </StyledForms>
  )
}

export default Forms
