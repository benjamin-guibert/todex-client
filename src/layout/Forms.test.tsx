import * as React from 'react'
import Forms from './Forms'
import renderer from 'react-test-renderer'

const renderComponent = () => renderer.create(<Forms />)

describe('<Forms />', () => {
  it('should render', () => {
    const result = renderComponent()

    expect(result.toJSON()).toMatchSnapshot()
  })
})
