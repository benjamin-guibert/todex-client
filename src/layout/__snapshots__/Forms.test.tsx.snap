// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Forms /> should render 1`] = `
<div
  className="sc-bdvvtL gdVBsa container"
>
  <div
    className="m-3 px-2 py-1 card"
  >
    <form
      className=""
    >
      <div
        className="my-3 row"
      >
        <div
          className="input-group has-validation"
        >
          <span
            className="input-group-text"
          >
            ETH
          </span>
          <input
            className="form-control"
            onChange={[Function]}
            placeholder="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="my-3 row"
      >
        <div
          className="btn-group"
          role="group"
        >
          <button
            className="btn btn-primary"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Deposit
          </button>
          <button
            className="btn btn-primary"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Withdraw
          </button>
        </div>
      </div>
    </form>
  </div>
  <div
    className="m-3 px-2 py-1 card"
  >
    <form
      className=""
    >
      <div
        className="my-3 row"
      >
        <div
          className="input-group has-validation"
        >
          <span
            className="input-group-text"
          >
            TDX
          </span>
          <input
            className="form-control"
            onChange={[Function]}
            placeholder="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="my-3 row"
      >
        <div
          className="btn-group"
          role="group"
        >
          <button
            className="btn btn-primary"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Deposit (Approve)
          </button>
          <button
            className="btn btn-primary"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Withdraw
          </button>
        </div>
      </div>
    </form>
  </div>
  <div
    className="m-3 px-2 py-1 text-dark card"
  >
    <form
      className=""
    >
      <div
        className="my-3 row"
      >
        <label
          className="text-end form-label col-form-label col-sm-3"
        >
          Amount
        </label>
        <div
          className="my-auto col"
        >
          <div
            className="input-group has-validation"
          >
            <span
              className="input-group-text"
            >
              TDX
            </span>
            <input
              className="form-control"
              onChange={[Function]}
              placeholder="0"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="my-3 row"
      >
        <label
          className="text-end form-label col-form-label col-sm-3"
        >
          Unit price
        </label>
        <div
          className="my-auto col"
        >
          <div
            className="input-group has-validation"
          >
            <span
              className="input-group-text"
            >
              ETH
            </span>
            <input
              className="form-control"
              onChange={[Function]}
              placeholder="0"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="my-3 row"
      >
        <label
          className="text-end form-label col-form-label col-sm-3"
        >
          Total price:
        </label>
        <div
          className="my-auto col"
        >
          <input
            className="font-monospace form-control-plaintext"
            defaultValue=""
            readOnly={true}
          />
        </div>
      </div>
      <div
        className="my-3 row"
      >
        <div
          className="btn-group"
          role="group"
        >
          <button
            className="btn btn-danger"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Buy
          </button>
          <button
            className="btn btn-success"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Sell
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
